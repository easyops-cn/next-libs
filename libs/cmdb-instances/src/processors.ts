import { getRuntime } from "@next-core/brick-kit";
import { uniq, compact } from "lodash";
import moment, { Moment } from "moment";
import { Query } from "@next-libs/cmdb-utils";

import { FormControlTypeEnum } from "./model-attribute-form-control/ModelAttributeFormControl";
import { clusterMap } from "./instance-list-table/constants";

const featureFlags =
  process.env.NODE_ENV === "test"
    ? { "cmdb-advance-search-with-quote": true }
    : getRuntime().getFeatureFlags();
export const ENABLED_CMDB_ADVANCE_SEARCH_WITH_QUOTE = true;
// featureFlags["cmdb-advance-search-with-quote"];

export const processAttrValueWithQuote = (
  attrValue: string,
  resultArray: any
): any => {
  const quoteIndex = attrValue.indexOf('"');
  const nextQuoteIndex = attrValue.indexOf('"', quoteIndex + 1);
  if (nextQuoteIndex == -1 || quoteIndex === -1) {
    if (attrValue.length) {
      resultArray = [...resultArray, ...attrValue.split(" ")];
    }
    return uniq(compact(resultArray));
  } else if (nextQuoteIndex != -1) {
    if (quoteIndex !== 0) {
      resultArray = [
        ...resultArray,
        ...attrValue.slice(0, quoteIndex).split(" "),
      ];
      resultArray.push(attrValue.slice(quoteIndex + 1, nextQuoteIndex));
      return processAttrValueWithQuote(
        attrValue.slice(nextQuoteIndex + 1),
        resultArray
      );
    } else {
      resultArray.push(attrValue.slice(quoteIndex + 1, nextQuoteIndex));
      return processAttrValueWithQuote(
        attrValue.slice(nextQuoteIndex + 1),
        resultArray
      );
    }
  }
};

export const computeDateFormat = (
  type: string,
  value: any
): { format: string; value: Moment | null } => {
  const dateFormat = "YYYY-MM-DD";
  const dateTimeFormat = "YYYY-MM-DD HH:mm:ss";
  let format;
  if (type === FormControlTypeEnum.DATETIME) {
    format = dateTimeFormat;
  } else {
    format = dateFormat;
  }
  return {
    value: value ? moment(value, format) : null,
    format,
  };
};

export const changeQueryWithCustomRules = (
  objectId: string,
  attrId: string,
  q: Query
): Query => {
  let modifiedQuery = q;
  if (objectId === "CLUSTER" && attrId === "type") {
    const operator = Object.keys(modifiedQuery["type"])[0];
    const value = Object.values(modifiedQuery["type"])[0];
    modifiedQuery = {
      type: {
        [operator]: clusterMap[value],
      },
    };
  }
  return modifiedQuery;
};

export const ATTRIBUTE_ID_PREFIX = "__ATTRIBUTE_ID_PREFIX__";
export const isClusterType = (objectId: string, type: string): boolean => {
  return (
    objectId === "CLUSTER" &&
    (type || "").replace(ATTRIBUTE_ID_PREFIX, "").replace("_", "") === "type"
  );
};
