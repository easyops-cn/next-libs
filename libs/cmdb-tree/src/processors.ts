import { isEmpty, cloneDeep, get } from "lodash";
import { CmdbModels } from "@next-sdk/cmdb-sdk";
import { TreeNode } from "./CMDBTree";

export function transformToTreeData(objectList: any[]): TreeNode[] {
  const categories = ["应用资源", "平台资源", "基础设施", "组织信息"];
  const map: any = { others: [] };
  for (const object of objectList) {
    object.key = object.objectId;
    object.title = object.name;
    object.type = "object";

    const category: string = object.category;
    if (category) {
      const tags = category.split(".", 2);
      if (categories.includes(tags[0])) {
        if (map[tags[0]] === undefined) {
          map[tags[0]] = {};
        }
        if (tags.length === 1) {
          tags.push("other-tag");
        }

        if (map[tags[0]][tags[1]] === undefined) {
          map[tags[0]][tags[1]] = [];
        }
        map[tags[0]][tags[1]].push(object);
      } else {
        if (map[category] === undefined) {
          map[category] = [];
        }
        map[category].push(object);
      }
    } else {
      map.others.push(object);
    }
  }

  const ret: TreeNode[] = [];
  for (const category of categories) {
    const children: TreeNode[] = [];
    if (map[category]) {
      for (const tag of Object.keys(map[category])) {
        if (tag !== "other-tag") {
          children.push({
            key: `${category}.${tag}`,
            title: tag,
            type: "tag",
            children: map[category][tag],
          });
        }
      }
      if (map[category]["other-tag"]) {
        children.push(...map[category]["other-tag"]);
      }
    }

    ret.push({
      key: category,
      title: category,
      type: "category",
      children,
    });
  }

  for (const key of Object.keys(map)) {
    if (categories.includes(key)) {
      continue;
    }

    ret.push(...map[key]);
  }
  return ret;
}

function match(treeNode: TreeNode, q: string): boolean {
  const str = q.toLowerCase();
  if (isEmpty(treeNode.children)) {
    if (
      treeNode.key.toLowerCase().includes(str) ||
      treeNode.title.toLowerCase().includes(str)
    ) {
      return true;
    }
    return false;
  }

  if (
    treeNode.key.toLowerCase().includes(str) ||
    treeNode.title.toLowerCase().includes(str)
  ) {
    return true;
  }
  treeNode.children = treeNode.children.filter((node) => match(node, q));
  return !isEmpty(treeNode.children);
}

export function search(treeData: TreeNode[], q: string): TreeNode[] {
  const nodes = cloneDeep(treeData);
  const result: TreeNode[] = [];
  for (const treeNode of nodes) {
    if (match(treeNode, q)) {
      result.push(treeNode);
    }
  }

  return result;
}

function getRelationObjectId(
  relations: Partial<CmdbModels.ModelObjectRelation>[],
  id: string
): string {
  let objectId = "";
  for (const relation of relations) {
    if (relation.left_id == id) {
      objectId = relation.right_object_id;
      break;
    } else if (relation.right_id == id) {
      objectId = relation.left_object_id;
      break;
    }
  }

  return objectId;
}

export function getObjectId2ShowKeys(
  objectList: CmdbModels.ModelCmdbObject[]
): Map<string, string[]> {
  const map = new Map();
  for (const model of objectList) {
    const showKeys = get(model, "view.show_key", ["name", "ip"]);
    map.set(model.objectId, showKeys);
  }
  return map;
}

export function fixRequestFields(
  objectList: CmdbModels.ModelCmdbObject[],
  request: CmdbModels.ModelInstanceTreeRootNode,
  notFixed: boolean,
  checkWhiteList: boolean
): string[] {
  const fields = new Set<string>();

  const objectId = request.object_id;
  const objectId2ShowKeys = getObjectId2ShowKeys(objectList);
  if (notFixed) {
    const reqFields = request.fields;
    if (reqFields) {
      for (const [key, value] of Object.entries(reqFields)) {
        if (value) {
          fields.add(key);
        }
      }
    }
  } else {
    const showKeys = objectId2ShowKeys.get(objectId);
    if (showKeys) {
      request.fields = {};
      if (checkWhiteList) {
        request.fields.readAuthorizers = true;
        request.fields.inheritedReadAuthorizers = true;
      }
      for (const showKey of showKeys) {
        request.fields[showKey] = true;
        fields.add(showKey);
      }
    } else {
      // eslint-disable-next-line no-console
      console.warn(`objectId '${objectId}' does NOT exist`);
    }
  }
  let map = new Map<string, CmdbModels.ModelInstanceTreeRootNode["child"]>();
  map.set(objectId, request.child);
  while (map.size > 0) {
    const childMap = new Map<
      string,
      CmdbModels.ModelInstanceTreeRootNode["child"]
    >();
    for (const [objectId, child] of map) {
      if (child?.length) {
        const model = objectList.find((o) => o.objectId == objectId);
        if (!model) {
          continue;
        }
        for (const c of child) {
          const id = getRelationObjectId(
            model.relation_list,
            c.relation_field_id
          );
          if (id) {
            childMap.set(id, c.child);
            if (notFixed) {
              const childFields = c.fields;
              if (childFields) {
                for (const [key, value] of Object.entries(childFields)) {
                  if (value) {
                    fields.add(key);
                  }
                }
              }
            } else {
              const showKeys = objectId2ShowKeys.get(id);
              c.fields = {};
              if (checkWhiteList) {
                c.fields.readAuthorizers = true;
                request.fields.inheritedReadAuthorizers = true;
              }
              for (const showKey of showKeys) {
                c.fields[showKey] = true;
                fields.add(showKey);
              }
            }
          } else {
            // eslint-disable-next-line no-console
            console.warn(
              `relation '${c.relation_field_id}' does NOT exist in '${objectId}'`
            );
          }
        }
      }
    }
    map = childMap;
  }

  return [...fields];
}

export function getTitle(
  instance: Record<string, any>,
  showKeys: string[]
): string {
  let title = get(instance, showKeys[0], "unknown");
  if (showKeys.length > 1 && get(instance, showKeys[1])) {
    title += `(${instance[showKeys[1]]})`;
  }
  return title;
}

export function getObjectIds(
  objectList: CmdbModels.ModelCmdbObject[],
  request: CmdbModels.ModelInstanceTreeRootNode
): string[] {
  const objectIds = new Set<string>();
  const objectId = request.object_id;
  let map = new Map<string, CmdbModels.ModelInstanceTreeRootNode["child"]>();
  map.set(objectId, request.child);
  objectIds.add(objectId);
  while (map.size > 0) {
    const childMap = new Map<
      string,
      CmdbModels.ModelInstanceTreeRootNode["child"]
    >();
    for (const [objectId, child] of map) {
      if (child?.length) {
        const model = objectList.find((o) => o.objectId == objectId);
        if (!model) {
          continue;
        }
        for (const c of child) {
          const id = getRelationObjectId(
            model.relation_list,
            c.relation_field_id
          );
          if (id) objectIds.add(id);
          childMap.set(id, c.child);
        }
      }
    }
    map = childMap;
  }

  return [...objectIds];
}

export function getRelation2ObjectId(
  objectList: CmdbModels.ModelCmdbObject[],
  request: CmdbModels.ModelInstanceTreeRootNode
): Map<string, string> {
  const ret = new Map<string, string>();
  const objectId = request.object_id;
  let map = new Map<string, CmdbModels.ModelInstanceTreeRootNode["child"]>();
  map.set(objectId, request.child);
  while (map.size > 0) {
    const childMap = new Map<
      string,
      CmdbModels.ModelInstanceTreeRootNode["child"]
    >();
    for (const [objectId, child] of map) {
      if (child?.length) {
        const model = objectList.find((o) => o.objectId == objectId);
        if (!model) {
          continue;
        }
        for (const c of child) {
          const id = getRelationObjectId(
            model.relation_list,
            c.relation_field_id
          );
          ret.set(c.relation_field_id, objectId);
          childMap.set(id, c.child);
        }
      }
    }
    map = childMap;
  }

  return ret;
}

export function updateChildren(
  instanceId: string,
  treeData: any,
  nodes: any[]
) {
  let data = treeData;
  while (data.length) {
    const children = [];
    for (const item of data) {
      if (item.key === instanceId) {
        item.children = nodes;
      }
      if (item.children) {
        children.push(...item.children);
      }
    }
    data = children;
  }
}
export function checkPermission(
  whiteList: string[],
  currentUser: string,
  userGroupIds: string[]
) {
  if (!isEmpty(whiteList)) {
    const userGroupIdSet = new Set(userGroupIds);
    return whiteList.some((i) => i === currentUser || userGroupIdSet.has(i));
  } else {
    return true;
  }
}
export function removeNoPermissionNode(tree: TreeNode[]): void {
  for (let i = tree.length - 1; i >= 0; i--) {
    const { authorized, children } = tree[i];
    if (children) {
      removeNoPermissionNode(children);
    }
    if (!authorized && !children?.length) {
      tree.splice(i, 1);
    }
  }
}
