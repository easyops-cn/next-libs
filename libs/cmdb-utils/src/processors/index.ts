export * from "./instance";
export * from "./object";
