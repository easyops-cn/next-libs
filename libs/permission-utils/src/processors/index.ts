export * from "./permissionCollection";
export * from "./permissionItem";
export * from "./processor";
