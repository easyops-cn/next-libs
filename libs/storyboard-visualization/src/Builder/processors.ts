import { sortBy, groupBy } from "lodash";
import { BuilderItem, BuilderNode } from "./interfaces";

export function builderToTree(list: BuilderItem[]): BuilderNode {
  if (list?.length === 1) {
    return builderItemToTree(list[0]);
  }
  return builderToTree([
    {
      alias: "APP",
      type: "app",
      children: list
    }
  ]);
}

const classNameMap = new Map<string, string>([
  ["brick", "brick"],
  ["provider", "brick"],
  ["template", "brick"],
  ["routes", "routes"],
  ["bricks", "routes"],
  ["redirect", "redirect"],
  ["app", "app"]
]);

export function getClassNameOfNodeType(type: string): string {
  if (classNameMap.has(type)) {
    return classNameMap.get(type);
  }
  return "node-type-unknown";
}

function builderItemToTree(data: BuilderItem): BuilderNode {
  const node: BuilderNode = {
    nodeData: data,
    groupIndex: 0,
    children: processChildren(data.children, shouldIgnoreGroup(data))
  };
  return node;
}

function shouldIgnoreGroup(data: BuilderItem): boolean {
  return ["routes", "bricks", "redirect", "app"].includes(data.type);
}

function processChildren(
  list: BuilderItem[],
  ignoreGroup?: boolean
): BuilderNode[] {
  const children: BuilderNode[] = [];
  if (Array.isArray(list)) {
    const sortedChildren = sortBy(list, ["sort"]);
    const groups = ignoreGroup
      ? sortedChildren.map(item => [item])
      : Object.values(groupBy(sortedChildren, "mountPoint"));
    let groupIndex = 0;
    for (const group of groups) {
      for (const child of group) {
        children.push({
          groupIndex,
          nodeData: child,
          children: processChildren(child.children, shouldIgnoreGroup(child))
        });
      }
      groupIndex += 1;
    }
  }
  return children;
}
