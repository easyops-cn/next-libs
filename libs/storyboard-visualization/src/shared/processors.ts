import { doTransform } from "@easyops/brick-kit";
import { ContentItemActions, ViewItem } from "./interfaces";
import { isObject } from "@easyops/brick-utils";
import { UseBrickConf } from "@easyops/brick-types";

export function filterActions(
  contentItemActions: ContentItemActions,
  item: ViewItem
): UseBrickConf[] {
  const filteredActions = []
    .concat(contentItemActions?.useBrick ?? [])
    .filter(action => {
      if (isObject(action.if)) {
        // eslint-disable-next-line
        console.warn("Currently don't support resolvable-if in `useBrick`");
      } else if (
        typeof action.if === "boolean" ||
        typeof action.if === "string"
      ) {
        const ifChecked = doTransform({ item }, action.if);
        if (ifChecked === false) {
          return false;
        }
      }
      return true;
    });
  return filteredActions;
}
