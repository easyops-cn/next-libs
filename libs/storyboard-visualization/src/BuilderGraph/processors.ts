import { sortBy, groupBy, isEmpty } from "lodash";
import {
  GraphNode,
  GraphNodeContent,
  GraphNodeContentSlotGroup,
} from "./interfaces";
import { ViewItem } from "../shared/interfaces";
import { styleConfig } from "./constants";
import { HierarchyPointLink } from "d3";

export function viewsToGraph(
  views: ViewItem[],
  wrapAnApp?: boolean | "auto"
): GraphNode {
  if (
    (wrapAnApp === false && views?.length >= 1) ||
    (wrapAnApp !== true && views?.length === 1)
  ) {
    if (views[0].type === "custom-template") {
      return ViewItemToGraph({
        alias: "Custom Template",
        type: "tpl-root",
        children: sortViews(views || []),
      });
    }
    return ViewItemToGraph(views[0]);
  }
  return viewsToGraph(
    [
      {
        alias: "APP",
        type: "app-root",
        children: sortViews(views || []),
      },
    ],
    false
  );
}

function ViewItemToGraph(view: ViewItem): GraphNode {
  const nodeType = isRouteNode(view)
    ? "route"
    : isBrickNode(view)
    ? "brick"
    : "unknown";
  const node: GraphNode = {
    nodeType,
    originalData: view,
    content: getNodeContent(view),
    children: sortViews(view.children)
      // Ignore nodes that has no children.
      .filter((child) => !isEmpty(child.children))
      .map((child) => ViewItemToGraph(child)),
  };
  node.height = computeNodeHeight(node);
  return node;
}

export function isRouteNode(view: ViewItem): boolean {
  return ["bricks", "routes", "redirect"].includes(view.type);
}

export function isBrickNode(view: ViewItem): boolean {
  return ["brick", "provider", "template", "custom-template"].includes(
    view.type
  );
}

function getNodeContent(view: ViewItem): GraphNodeContent {
  switch (view.type) {
    case "bricks":
    case "routes":
    case "app-root":
    case "tpl-root":
    case "custom-template":
      return {
        type:
          view.type === "app-root"
            ? "routes"
            : view.type === "tpl-root"
            ? "custom-template"
            : view.type,
        items: sortViews(view.children),
      };
    case "redirect":
      return {
        type: "redirect",
      };
    case "brick":
      return {
        type: "slots",
        slots: getNodeContentSlotGroups(view),
      };
    default:
      return;
  }
}

function getNodeContentSlotGroups(view: ViewItem): GraphNodeContentSlotGroup[] {
  let slotGroups: GraphNodeContentSlotGroup[] = [];
  if (Array.isArray(view.children)) {
    const sortedChildren = sortViews(view.children);
    slotGroups = Object.entries(
      groupBy(sortedChildren, (item) => item.mountPoint ?? "")
    ).map(([name, items]) => ({
      name,
      type: isRouteNode(items[0])
        ? "routes"
        : isBrickNode(items[0])
        ? "bricks"
        : "unknown",
      items,
    }));
  }
  return slotGroups;
}

export function getNodeDisplayName(data: ViewItem): string {
  if (data.alias) {
    return data.alias;
  }

  if (data.type === "custom-template") {
    return data.templateId;
  }

  if (isBrickNode(data)) {
    return getBrickName(data);
  }

  return data.path;
}

function getBrickName(data: ViewItem, showFullBrickName?: boolean): string {
  const brickName = data.template || data.brick;
  if (showFullBrickName) {
    return brickName;
  }
  return brickName.split(".").slice(-1)[0];
}

function computeNodeHeight(node: GraphNode): number {
  let height = styleConfig.node.padding * 2 + styleConfig.alias.height;
  if (node.content) {
    switch (node.content.type) {
      case "bricks":
      case "routes":
      case "custom-template":
        height +=
          node.content.items.length * styleConfig.contentItem.height +
          (node.content.items.length - 1) *
            styleConfig.contentItem.marginBottom;
        break;
      case "slots":
        height +=
          node.content.slots.reduce(
            (acc, group) =>
              acc +
              styleConfig.contentDivider.height +
              group.items.length * styleConfig.contentItem.height +
              (group.items.length - 1) * styleConfig.contentItem.marginBottom,
            0
          ) +
          (node.content.slots.length - 1) *
            styleConfig.contentGroup.marginBottom;
        break;
    }
  }
  return height;
}

export function computeSourceX({
  source,
  target,
}: HierarchyPointLink<GraphNode>): number {
  const content = source.data.content;

  if (!content) {
    return source.x;
  }

  let x =
    source.x -
    source.data.height / 2 +
    styleConfig.node.padding +
    styleConfig.alias.height;

  let item: ViewItem;
  let slot: GraphNodeContentSlotGroup;

  switch (content.type) {
    case "bricks":
    case "routes":
    case "custom-template":
      for (item of content.items) {
        if (item === target.data.originalData) {
          return x + styleConfig.contentItem.height / 2;
        }
        x +=
          styleConfig.contentItem.height + styleConfig.contentItem.marginBottom;
      }
      // Should never reach `break`.
      /* istanbul ignore next */
      break;
    case "slots":
      for (slot of content.slots) {
        x += styleConfig.contentDivider.height;
        for (item of slot.items) {
          if (item === target.data.originalData) {
            return x + styleConfig.contentItem.height / 2;
          }
          x +=
            styleConfig.contentItem.height +
            styleConfig.contentItem.marginBottom;
        }
        x -= styleConfig.contentItem.marginBottom;
        x += styleConfig.contentGroup.marginBottom;
      }
      // Should never reach `break`.
      /* istanbul ignore next */
      break;
  }

  return source.x;
}

function sortViews(views: ViewItem[]): ViewItem[] {
  return sortBy(views || [], [(item) => item.sort ?? -Infinity]);
}
