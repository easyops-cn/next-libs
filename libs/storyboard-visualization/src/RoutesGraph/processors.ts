import { ViewItem } from "../shared/interfaces";
import { GraphNode } from "./interfaces";
import { set, uniqueId } from "lodash";

const mockTarget = {
  "0": [1, 2],
  "1": [3]
};

const mockData = {
  0: {
    x: 300,
    y: 10,
    viewType: "dashboard"
  },
  1: {
    x: 150,
    y: 360,
    viewType: "mixed"
  },
  2: {
    x: 450,
    y: 360,
    viewType: "tabs"
  },
  3: {
    x: 300,
    y: 710,
    viewType: "tableList"
  }
};

export function viewsToGraph(views: ViewItem[]): GraphNode[] {
  return views?.map((item, i) => {
    if (mockTarget[i]) {
      mockTarget[i].forEach(target => {
        set(item, `segues.${uniqueId("go-")}.target`, views[target].alias);
      });
    }
    item.x = mockData[i].x;
    item.y = mockData[i].y;
    item.viewType = mockData[i].viewType;
    return {
      originalData: item
    };
  });
}
