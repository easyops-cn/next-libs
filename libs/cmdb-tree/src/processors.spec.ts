import { transformToTreeData, search } from "./processors";

const objectList = [
  {
    category: "",
    objectId: "HELLO",
    name: "hello"
  },
  {
    category: "基础设施.F5",
    objectId: "iRoute",
    name: "iRoute"
  },
  {
    category: "应用资源",
    objectId: "BUSINESS",
    name: "业务"
  },
  {
    category: "应用资源",
    objectId: "APP",
    name: "应用"
  },
  {
    category: "应用资源.部署资源",
    objectId: "RESOURCE_GROUP",
    name: "部署资源组"
  },
  {
    category: "others.foo",
    objectId: "OTHERS",
    name: "others"
  },
  {
    category: "others.foo",
    objectId: "xOTHERS",
    name: "others"
  }
];

describe("transformToTreeData", () => {
  it("transformToTreeData should work", () => {
    const result = transformToTreeData(objectList);
    expect(result).toMatchSnapshot();
  });

  it("search should work", () => {
    const treeData = [
      {
        key: "id-1",
        title: "name-1",
        children: [
          { key: "id-1-1", title: "name-1-1" },
          { key: "id-1-2", title: "name-1-2" },
          { key: "id-1-3", title: "name-1-3" }
        ]
      },
      { key: "id-2", title: "name-2" },
      { key: "id-3", title: "name-3" }
    ];

    let result = search(treeData, "id-1-3");
    expect(result).toEqual([
      {
        key: "id-1",
        title: "name-1",
        children: [{ key: "id-1-3", title: "name-1-3" }]
      }
    ]);

    result = search(treeData, "name-1-2");
    expect(result).toEqual([
      {
        key: "id-1",
        title: "name-1",
        children: [{ key: "id-1-2", title: "name-1-2" }]
      }
    ]);

    expect(search(treeData, "xxx").length).toBe(0);

    result = search(treeData, "name-1");
    expect(result[0]).toEqual(treeData[0]);
  });
});
