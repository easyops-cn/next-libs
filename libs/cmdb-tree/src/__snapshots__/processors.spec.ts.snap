// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformToTreeData transformToTreeData should work 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "category": "应用资源.部署资源",
            "key": "RESOURCE_GROUP",
            "name": "部署资源组",
            "objectId": "RESOURCE_GROUP",
            "title": "部署资源组",
            "type": "object",
          },
        ],
        "key": "应用资源.部署资源",
        "title": "部署资源",
        "type": "tag",
      },
      Object {
        "category": "应用资源",
        "key": "BUSINESS",
        "name": "业务",
        "objectId": "BUSINESS",
        "title": "业务",
        "type": "object",
      },
      Object {
        "category": "应用资源",
        "key": "APP",
        "name": "应用",
        "objectId": "APP",
        "title": "应用",
        "type": "object",
      },
    ],
    "key": "应用资源",
    "title": "应用资源",
    "type": "category",
  },
  Object {
    "children": Array [],
    "key": "平台资源",
    "title": "平台资源",
    "type": "category",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "category": "基础设施.F5",
            "key": "iRoute",
            "name": "iRoute",
            "objectId": "iRoute",
            "title": "iRoute",
            "type": "object",
          },
        ],
        "key": "基础设施.F5",
        "title": "F5",
        "type": "tag",
      },
    ],
    "key": "基础设施",
    "title": "基础设施",
    "type": "category",
  },
  Object {
    "children": Array [],
    "key": "组织信息",
    "title": "组织信息",
    "type": "category",
  },
  Object {
    "category": "",
    "key": "HELLO",
    "name": "hello",
    "objectId": "HELLO",
    "title": "hello",
    "type": "object",
  },
  Object {
    "category": "others.foo",
    "key": "OTHERS",
    "name": "others",
    "objectId": "OTHERS",
    "title": "others",
    "type": "object",
  },
  Object {
    "category": "others.foo",
    "key": "xOTHERS",
    "name": "others",
    "objectId": "xOTHERS",
    "title": "others",
    "type": "object",
  },
]
`;
